rfcs/0112-demote-darwin-support-tier.md

```diff
@@ -0,0 +1,91 @@
+shepherd-leader: (name to be appointed by RFC steering committee)
+related-issues: (will contain links to implementation PRs)
+---
+
+# Summary
+[summary]: #summary
+
+Move x86_64-darwin from a Tier 2 platform to Tier 3, as described in [RFC 0046](https://github.com/NixOS/rfcs/blob/master/rfcs/0046-platform-support-tiers.md).
+
+# Motivation
+[motivation]: #motivation
+
+Darwin is constantly adding additional maintainer burden. Especially, it does not live up to the requirements:
+
+- "A lot of packages built by Hydra, full ofBorg support."
+    - There is only one ofBorg builder. It can take several hours or days for an ofBorg build to finish for Darwin, if it finishes at all.
+    - Hydra builders a greatly understaffed.
+    - If Apple follows through on [their announcement to discontinue](https://www.businessinsider.com/apple-macbook-pro-discontinued) x86-based devices, the situation won't improve due to a physical lack of hardware.
+- "Most packages work, credible ambition to reach Tier 1 coverage at some point."
+    - Tier 1 support is far out of reach.
+    - We lack expertise in the greater community to support Darwin.
+- "Some ordinary packages are channel blockers on Hydra."
+    - Due to Hydra being understaffed, this is a problem (see above).
+
+We are not only lacking CI machines, but also maintainers that can dedicate their time to work on fixing broken Darwin packages.
+
+## Exemplary collection of issues with Darwin
+
+The issues that are constantly encountered in relation to the Darwin platform have a genuine negative impact on the `nixpkgs` development experience. People are forced to deal with these problems even though they have no hardware to test on, lack the proper expertise, and are reluctant when it comes to investing additional time and energy to patch for the failing platform. Demoting Darwin to Tier 3 will allow more developers to refocus their attention on better supported platforms, care less about Darwin issues, and push that burden onto the Darwin maintainers.
+
+An excerpt of these issues are:
+
+- The only way to debug issues without owning a Mac is to run Darwin in a VM, of which the legal situation is unclear.
+   - Those licensing requirements are causing friction with our CI setup as well.
+- The sandboxing capabilities on MacOS are limited. This results in a special category of test failures only relevant to Darwin. It also prevents us from providing a community builder, because it could not be shared between multiple people that cannot trust each other.
+- The MacOS SDK cannot be updated and is stuck on 10.12 (released 2016/09) because Apple does not publish the required sources.
+    - https://github.com/NixOS/nixpkgs/issues/101229#issuecomment-938747052
+    - It is highly unlikely that we can report bugs to projects with such an outdated and unsupported SDK version
+- The MacOS SDK being outdated blocks nixpkgs moving to go 1.17:
```

Comment:
From the September entry:

> As for the SDK bump, configd remains a hard nut to crack. Somewhere around macOS 10.13 Apple has stopped releasing all the headers that are necessary to build it. XPC no longer seems avoidable as a dependency, CoreFoundation seems incomplete (we rely on Darling for the missing bits now), there’s some packages hosted on opensource.apple.com 3 but not listed in the macOS releases, like neon and OpenBSM. I don’t see an alternative to getting the XPC headers from the SDK and likewise for some other missing headers. There’s several projects online where missing headers are worked around by stubbing, like OSXPrivateSDK 3 and GoVPN 3, but it doesn’t seem like a good way to deal with the issue. Fabricating a constant can lead to unexpected behavior. That’s why I intend to use binaries from the SDK whenever dependencies aren’t available. This will be a step back with regards to building from source unfortunately.

That doesn’t sound to me like it’s going to get better in the future, or whether it’s even doable in the first place.

Reply:
It's very much doable, we just don't know how much of the stdenv will contain binary blobs. That's the case with darwin anyway, since it's linking against system frameworks and libSystem.

@toonn is working on this and it's indeed a lot of work, but it will get better :)